src/service/fixed-layer.js

```diff
@@ -100,27 +104,89 @@ export class FixedLayer {
 
     /** @private {?MutationObserver} */
     this.mutationObserver_ = null;
+
+    /** @private {?Array<string>} */
+    this.fixedSelectors_ = null;
+
+    /** @private {?Array<string>} */
+    this.stickySelectors_ = null;
   }
 
   /**
-   * @param {boolean} visible
+   * Informs FixedLayer that a lightbox was opened.
+   *
+   * - FixedLayer hides any transfer layer elements that may be overlayed on
+   *   top of the lightbox, which is confusing UX. When `onComplete` resolves,
+   * - FixedLayer scans and transfers any fixed descendants of `lightbox`.
+   *   This enables unjanky fixed elements in lightboxes on iOS.
+   *
+   * @param {!Element=} opt_lightbox
+   * @param {!Promise=} opt_onComplete Promise that resolves when lightbox
+   *   UX completes e.g. open transition animation.
    */
-  setVisible(visible) {
-    if (this.transferLayer_) {
-      this.vsync_.mutate(() => {
-        setStyle(
-            this.transferLayer_.getRoot(),
-            'visibility',
-            visible ? 'visible' : 'hidden');
+  enterLightbox(opt_lightbox, opt_onComplete) {
+    const isLightboxExperimentOn =
+        isExperimentOn(this.ampdoc.win, 'fixed-elements-in-lightbox');
+
+    this.setLightboxMode_(true, isLightboxExperimentOn);
+
+    if (isLightboxExperimentOn && opt_lightbox && opt_onComplete) {
+      opt_onComplete.then(() => {
+        this.scanNode_(dev().assertElement(opt_lightbox),
+            /* lightboxMode */ true);
       });
     }
   }
 
+  /**
+   * Reverses the actions performed by `enterLightbox()`.
+   */
+  leaveLightbox() {
+    const isLightboxExperimentOn =
+        isExperimentOn(this.ampdoc.win, 'fixed-elements-in-lightbox');
+
+    this.setLightboxMode_(false, isLightboxExperimentOn);
+
+    if (isLightboxExperimentOn) {
+      const fes = remove(this.elements_, fe => !!fe.lightboxed);
+      this.returnFixedElements_(fes);
+      if (!this.elements_.length) {
```

Comment:
Why move this out of `returnFixedElements_`?

Reply:
Didn't feel like it belongs in a function called `returnFixedElements_`, which I renamed because I cringed after looking at `removeElement` and `removeElement_`. 

Also did you notice "element" refers to both `Element` and `ElementDef` everywhere?

Reply:
Move this to a specific method, then.

Yah, blame Dima. 😛